cirq-google/cirq_google/workflow/quantum_runtime.py

```diff
@@ -276,10 +277,9 @@ def execute(
     _update_updatable_files(egr_record, shared_rt_info, data_dir)
     executable_results = []
 
-    # Loop over executables.
     sampler = rt_config.processor.get_sampler()
-    n_executables = len(executable_group)
-    print()
+    logger = _PrintLogger(n_total=len(executable_group))
+    logger.initialize()
```

Comment:
With this PR and PR #4614, there is a common pattern here of creating the object, initializing it, and then calling a method on it in the loop. What do we think of having something like this:

```python3
def execute(..., callbacks): # list of callback classes you want to have for this execution.
  .
  .
  .

  for callback in callbacks:
    callback.initialize()

  for i, exe in enumerate(executable_group):
    .
    .
    .
    for callback in callbacks:
      callback.update_step(exe_result, shared_rt_info, any other relevant metadata/ args leaning towards adding more)

  for callback in ballcks:
    callback.finalize()
```

Then you can make a SaverCallback class who's __init__ is the desired dir to save stuff and it defines a `intialize`, `finalize()` and `update_step`. PrintLogger can also follow this same structure too. This would make adding more callbacks in the future much simpler. WDYT ?

Reply:
good question. let me think about it

Reply:
I think I'd be ok moving this paradigm for these "classical" questions of how to save data and show progress. For any of the forthcoming quantum-pipeline features we might need more specialized APIs.

Are you ok merging this and #4614 and doing the callback factoring in a follow-on PR. Since these PRs don't actually alter the user-facing API at all (yet) it should work out great.

Reply:
SGTM